pandas/core/indexes/base.py

```diff
@@ -177,10 +179,15 @@ class Index(IndexOpsMixin, PandasObject):
     str = accessor.AccessorProperty(strings.StringMethods)
 
     def __new__(cls, data=None, dtype=None, copy=False, name=None,
-                fastpath=False, tupleize_cols=True, **kwargs):
+                fastpath=False, tupleize_cols=True, names=None,
+                **kwargs):
 
-        if name is None and hasattr(data, 'name'):
-            name = data.name
+        if name is None:
+            if hasattr(data, 'name'):
+                name = data.name
+            # extract `name` from `names` in case MultiIndex cannot be created
+            elif names:
+                name = names[0]
```

Comment:
add an 

```
elif names is not None:
     if name is not None:
           raise ValuError(....)
    name = names
```

Reply:
I think, it's redundant as this piece of code is inside `if name is None` block.
Maybe it's better to raise an error in this case before other checks without nesting. And we still need `name` to be a single object.
What about the solution below?
```python
if names is not None and name is not None:
    raise TypeError("Can only provide one of `names` and `name`")

if name is None:
    if hasattr(data, 'name'):
        name = data.name
    # extract `name` from `names` in case MultiIndex cannot be created
    elif names:
        name = names[0] if is_list_like(names) else names
```

Reply:
I think, it's redundant as this piece of code is inside `if name is None` block.
Maybe it's better to raise an error in this case before other checks without nesting. And we still need `name` to be a single object.
What about the solution below?
```python
if names is not None and name is not None:
    raise TypeError("Can only provide one of `names` and `name`")

if name is None:
    if hasattr(data, 'name'):
        name = data.name
    # extract `name` from `names` in case MultiIndex cannot be created
    elif names:
        name = names[0] if is_list_like(names) else names
```